File: src/io/socket_manager.py
import socket
from PySide6.QtCore import Signal
from src.io.io_transport import IOTransport
from src.io.socket_reader import SocketReaderThread


def get_local_ips():
    """获取本机所有 IPv4 地址"""
    ips = []
    try:
        hostname = socket.gethostname()
        for info in socket.getaddrinfo(hostname, None):
            addr = info[4][0]
            if addr and not addr.startswith('127.') and '.' in addr:
                ips.append(addr)
    except Exception:
        pass
    ips.append('0.0.0.0')
    ips.append('127.0.0.1')
    seen = set()
    return [x for x in ips if not (x in seen or seen.add(x))]


class SocketManager(IOTransport):
    client_event = Signal(str, tuple)

    def __init__(self):
        super().__init__()
        self.sock = None
        self.server_sock = None
        self._mode = None
        self._remote_addr = None
        self.settings = {
            'host': '0.0.0.0',
            'port': 8080,
            'protocol': 'TCP',
            'role': 'Server',
            'frame_timeout': 50,
        }

    def get_available_devices(self):
        return get_local_ips()

    @property
    def current_client(self):
        if self.reader_thread:
            return self.reader_thread.current_client
        return None

    @property
    def mode(self):
        return self._mode

    def open_connection(self, host, port, protocol='TCP', role='Client'):
        return super().open_connection(host, port, protocol=protocol, role=role)

    def _connect_impl(self, host, port, protocol='TCP', role='Client'):
        self._open_socket(host, port, protocol, role)
        frame_timeout = self.settings.get('frame_timeout', 50)
        thread = SocketReaderThread(
            self.server_sock if self._mode == 'tcp_server' else self.sock,
            self._mode, frame_timeout,
        )
        thread.client_event.connect(self.client_event)
        self._connect_reader_thread(thread)

    def _open_socket(self, host, port, protocol, role):
        """创建并配置底层 socket，设置 self._mode。"""
        if protocol == 'TCP' and role == 'Server':
            self.server_sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            self.server_sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
            self.server_sock.setblocking(False)
            self.server_sock.bind((host, port))
            self.server_sock.listen(5)
            self._mode = 'tcp_server'

        elif protocol == 'TCP' and role == 'Client':
            self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            self.sock.settimeout(5)
            self.sock.connect((host, port))
            self.sock.setblocking(False)
            self._mode = 'tcp_client'

        elif protocol == 'UDP' and role == 'Server':
            self.sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
            self.sock.setblocking(False)
            self.sock.bind((host, port))
            self._mode = 'udp_server'

        elif protocol == 'UDP' and role == 'Client':
            self.sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
            self.sock.setblocking(False)
            self._remote_addr = (host, port)
            self._mode = 'udp_client'

        else:
            raise ValueError(f"Unknown mode: {protocol}/{role}")

    def _close_resource(self):
        if self.server_sock:
            try:
                self.server_sock.close()
            except Exception:
                pass
            self.server_sock = None
        if self.sock:
            try:
                self.sock.close()
            except Exception:
                pass
            self.sock = None
        self._remote_addr = None
        self._mode = None

    def _send_bytes(self, data: bytes) -> bool:
        if self._mode == 'tcp_client':
            self.sock.send(data)
        elif self._mode == 'tcp_server':
            if not self.reader_thread or not self.reader_thread.send_to_current(data):
                self.error_occurred.emit("没有已连接的客户端")
                return False
        elif self._mode in ('udp_server', 'udp_client'):
            if self._mode == 'udp_client':
                self.sock.sendto(data, self._remote_addr)
            else:
                if self.reader_thread and self.reader_thread.current_client:
                    self.sock.sendto(data, self.reader_thread.current_client)
                else:
                    self.error_occurred.emit("没有客户端地址")
                    return False
        return True

File: src/io/socket_manager.cpp
#include "socket_manager.h"
#include "socket_reader.h"

#include <QNetworkInterface>
#include <QHostAddress>

SocketManager::SocketManager(QObject *parent)
    : IOTransport(parent)
{
}

QStringList SocketManager::getLocalIPs() {
    QStringList ips;
    ips << "0.0.0.0" << "127.0.0.1";

    for (const QNetworkInterface &iface : QNetworkInterface::allInterfaces()) {
        if (!(iface.flags() & QNetworkInterface::IsUp)) continue;
        if (iface.flags() & QNetworkInterface::IsLoopBack) continue;

        for (const QNetworkAddressEntry &entry : iface.addressEntries()) {
            if (entry.ip().protocol() == QAbstractSocket::IPv4Protocol) {
                QString ip = entry.ip().toString();
                if (!ip.startsWith("127.") && !ips.contains(ip)) {
                    ips << ip;
                }
            }
        }
    }
    return ips;
}

QList<QPair<QString, QString>> SocketManager::getAvailableDevices() {
    QList<QPair<QString, QString>> devices;
    for (const QString &ip : getLocalIPs()) {
        devices.append({ip, ip});
    }
    return devices;
}

QPair<QString, int> SocketManager::currentClient() const {
    return m_currentClient;
}

bool SocketManager::connectImpl(const QVariantMap &params) {
    QString host = params.value("host", "127.0.0.1").toString();
    int port = params.value("port", 8080).toInt();
    QString protocol = params.value("protocol", "tcp").toString();
    QString role = params.value("role", "client").toString();

    return openSocket(host, port, protocol, role);
}

bool SocketManager::openSocket(const QString &host, int port,
                                const QString &protocol, const QString &role) {
    closeResource();

    if (protocol == "tcp") {
        if (role == "server") {
            m_mode = "tcp_server";
            m_tcpServer = new QTcpServer(this);

            if (!m_tcpServer->listen(QHostAddress::Any, port)) {
                emit errorOccurred("TCP Server listen failed: " + m_tcpServer->errorString());
                return false;
            }

            // Create reader thread for server
            auto *reader = new SocketReaderThread(m_tcpServer, m_mode, this);
            connect(reader, &SocketReaderThread::clientEvent,
                    this, &SocketManager::clientEvent);
            connect(reader, &SocketReaderThread::dataReceived, this, &IOTransport::dataReceived);
            connect(reader, &SocketReaderThread::errorOccurred, this, &IOTransport::errorOccurred);
            startReaderThread(reader);

            return true;

        } else {
            m_mode = "tcp_client";
            m_tcpSocket = new QTcpSocket(this);
            m_tcpSocket->connectToHost(host, port);

            if (!m_tcpSocket->waitForConnected(5000)) {
                emit errorOccurred("TCP connect failed: " + m_tcpSocket->errorString());
                return false;
            }

            // Create reader thread for client
            auto *reader = new SocketReaderThread(m_tcpSocket, m_mode, this);
            connect(reader, &SocketReaderThread::dataReceived, this, &IOTransport::dataReceived);
            connect(reader, &SocketReaderThread::errorOccurred, this, &IOTransport::errorOccurred);
            startReaderThread(reader);

            return true;
        }
    } else { // UDP
        m_mode = (role == "server") ? "udp_server" : "udp_client";
        m_udpSocket = new QUdpSocket(this);

        if (m_mode == "udp_server") {
            if (!m_udpSocket->bind(QHostAddress::Any, port)) {
                emit errorOccurred("UDP bind failed");
                return false;
            }
        } else {
            m_remoteAddr = {host, port};
        }

        // Create reader thread for UDP
        auto *reader = new SocketReaderThread(m_udpSocket, m_mode, this);
        connect(reader, &SocketReaderThread::dataReceived, this, &IOTransport::dataReceived);
        connect(reader, &SocketReaderThread::errorOccurred, this, &IOTransport::errorOccurred);
        startReaderThread(reader);

        return true;
    }
}

void SocketManager::closeResource() {
    if (m_tcpServer) {
        m_tcpServer->close();
        delete m_tcpServer;
        m_tcpServer = nullptr;
    }
    if (m_tcpSocket) {
        m_tcpSocket->disconnectFromHost();
        if (m_tcpSocket->state() != QAbstractSocket::UnconnectedState) {
            m_tcpSocket->waitForDisconnected(1000);
        }
        delete m_tcpSocket;
        m_tcpSocket = nullptr;
    }
    if (m_udpSocket) {
        m_udpSocket->close();
        delete m_udpSocket;
        m_udpSocket = nullptr;
    }
    m_mode.clear();
}

bool SocketManager::sendBytes(const QByteArray &data) {
    if (m_mode == "tcp_client") {
        if (!m_tcpSocket) return false;
        return m_tcpSocket->write(data) == data.size();

    } else if (m_mode == "tcp_server") {
        // Delegate to reader thread's send_to_current
        // For now, find the reader thread and call sendToCurrent
        if (m_readerThread) {
            auto *reader = qobject_cast<SocketReaderThread *>(m_readerThread);
            if (reader) {
                return reader->sendToCurrent(data);
            }
        }
        return false;

    } else if (m_mode == "udp_client") {
        if (!m_udpSocket) return false;
        qint64 sent = m_udpSocket->writeDatagram(data,
            QHostAddress(m_remoteAddr.first), m_remoteAddr.second);
        return sent == data.size();

    } else if (m_mode == "udp_server") {
        if (!m_udpSocket) return false;
        qint64 sent = m_udpSocket->writeDatagram(data,
            QHostAddress(m_currentClient.first), m_currentClient.second);
        return sent == data.size();
    }

    return false;
}
